core/src/main/scala/org/apache/spark/crypto/CryptoCodec.scala

```
@@ -0,0 +1,144 @@
+   * a lossless operation (concatenation, addition, or XOR).
+   * @see http://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Counter_.28CTR.29
+   * @param initIV
+   * @param counter
+   * @param IV
+   */
+  def calculateIV(initIV: Array[Byte], counter: Long, IV: Array[Byte])
+
+  /**
+   * @return Encryptor the encryptor
+   */
+  def createEncryptor: Encryptor
+
+  /**
+   * @return Decryptor the decryptor
+   */
+  def createDecryptor: Decryptor
+
+  /**
+   * Generate a number of secure, random bytes suitable for cryptographic use.
+   * This method needs to be thread-safe.
+   * @param bytes byte array to populate with random data
+   */
+  def generateSecureRandom(bytes: Array[Byte])
+}
+
+object CryptoCodec extends Logging {
+  def getInstance(conf: SparkConf): CryptoCodec = {
+    var name: String = conf.get(SPARK_SECURITY_CRYPTO_CIPHER_SUITE_KEY,
+      SPARK_SECURITY_CRYPTO_CIPHER_SUITE_DEFAULT)
+    getInstance(conf, CipherSuite.convert(name))
+  }
+
+  def getInstance(conf: SparkConf, cipherSuite: CipherSuite): CryptoCodec = {
+    var klasses: List[String] = getCodecClasses(conf, cipherSuite)
+    var codec: CryptoCodec = null
+    for (klass <- klasses) {
+      try {
+        val m = universe.runtimeMirror(getClass.getClassLoader)
+        var c: CryptoCodec = null
+        if (klass.equals("org.apache.spark.crypto.JceAesCtrCryptoCodec")) {
+          val classCryptoCodec = universe.typeOf[org.apache.spark.crypto.JceAesCtrCryptoCodec]
+            .typeSymbol.asClass
+          val cm = m.reflectClass(classCryptoCodec)
+          val ctor = universe.typeOf[org.apache.spark.crypto.JceAesCtrCryptoCodec].declaration(
+            universe.nme.CONSTRUCTOR).asMethod
+          val ctorm = cm.reflectConstructor(ctor)
+          val p = ctorm(conf)
+          c = p.asInstanceOf[org.apache.spark.crypto.CryptoCodec]
+        } else {
+          // TODO add Openssl crypto codec
```

Should throw an exception.